### ISSUE TITLE
ReduceLROnPlateau parent class is not _LRScheduler

### ISSUE THREAD

**Issue by anjandeepsahni** [pytorch/pytorch#21981]
## 🐛 Bug

This is based on the discussion here: [forum link](https://discuss.pytorch.org/t/reducelronplateau-parent-class-is-not-lrscheduler/48441)

I was wondering why ReduceLROnPlateau does not have _LRScheduler as parent class. If I want to check if a variable is a learning rate scheduler object type, I could have used:

`isinstance(scheduler, optim.lr_scheduler._LRScheduler)`

Above covers all learning rate schedulers defined in lr_scheduler.py except ReduceLROnPlateau because it does not inherit from _LRScheduler. So instead I need to use:

`isinstance(scheduler, optim.lr_scheduler._LRScheduler) or isinstance(scheduler, optim.lr_scheduler.ReduceLROnPlateau)`

Shouldn’t ReduceLROnPlateau inherit from _LRScheduler as well? Is this just an oversight?

Context:
In general for saving a checkpoint, if I also want to save scheduler state dict, then I would like to verify the params first.

## To Reproduce

```python
import torch
import torchvision
import torch.optim as optim
model = torchvision.models.alexnet(pretrained=False)
optimizer = optim.SGD(model.parameters(), lr=0.5, momentum=0.9, weight_decay=0.5)
scheduler = optim.lr_scheduler.ReduceLROnPlateau(optimizer, patience=4, threshold=0.01, verbose=True)

isinstance(scheduler, optim.lr_scheduler._LRScheduler)
# `out: False`

isinstance(scheduler, optim.lr_scheduler.ReduceLROnPlateau)
# `out: True`
```

## Environment

PyTorch version: 1.1.0
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 16.04.4 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 1080 Ti
Nvidia driver version: 418.56
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy==1.13.3
[pip] numpydoc==0.7.0
[pip] torch==1.1.0
[pip] torchvision==0.3.0
[conda] blas                      1.0                         mkl  
[conda] mkl                       2018.0.0             hb491cac_4  
[conda] mkl-service               1.1.2            py36h17a0993_4  
[conda] pytorch                   1.1.0           py3.6_cuda9.0.176_cudnn7.5.1_0    pytorch
[conda] torchvision               0.3.0           py36_cu9.0.176_1    pytorch

**Comment by SsnL** [comment]
I did not look into how possible it is to make it a subclass of `_LRScheduler`. If it isn't hard, we could do it. 

However, you should never rely on private classes. If any user code is broken by assuming non public API, we may not fix.

**Comment by VitalyFedyunin** [comment]
`ReduceLROnPlateau` overrides every `_LRScheduler` function. We never had need to have `_LRScheduler` as parent. 

Plus, as already mentioned, usage of any internal API names is a terrible idea. 

I will close the issue as non-bug. However feel free to send us PR which will introduce good method of checking if object is Scheduler

**Comment by anjandeepsahni** [comment]
Thanks for clarifying.

**Comment by ForrestCKoch** [comment]
> `ReduceLROnPlateau` overrides every `_LRScheduler` function. We never had need to have `_LRScheduler` as parent.

It doesn't seem to implement the `get_lr` function specified by the `_LRScheduler` class actually.  I agree with the sentiment of @anjandeepsahni that it would be nice to be able to rely on LR Schedulers to inherit from a common interface.  The above mentioned `get_lr` problem as well as the changed `step` function (which requires a `float` to be passed to it) may cause difficulty with code that allows different schedulers to be plugged in.

**Comment by jesford** [comment]
I realize this issue has been closed and understand that relying on internal API names is not ideal. However, I do agree with the other comments here, that it would be nice to inherit from a common class, consistent with every single other `lr_scheduler`. Would you accept a PR that makes `ReduceLROnPlateau` inherit from `_LRScheduler`?

**Comment by VitalyFedyunin** [comment]
cc @vincentqb do we have someone to implement it, or we should wait for open source PR

**Comment by vincentqb** [comment]
I would accept a PR to do this as long as it does not change the functionality.
